src/Renci.SshNet/Common/Pipe.cs

```diff
@@ -0,0 +1,47 @@
+﻿namespace Renci.SshNet.Common
+{
+    using System;
+    using System.IO;
+
+    /// <summary>
+    /// A generic pipe to pass through data.
+    /// </summary>
+    internal class Pipe : IDisposable
+    {
```

Comment:
What if `EndExecute` **Result** and **Error** or `Execute` return a result object instead that has those properties?

In general it doesn't sound like a good idea to have them exposed until execution has ended with any other mean than through the streams.

On way could be that `Execute` would be a shortcut that reads both pipes to EOS and returns a result object with both where the `BeginExecute` and `EndExecute` way could only expose the streams?